File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/1300-K 번째 수/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

int N, K;

void input() {
	cin >> N >> K;
}

bool determination(ll candidate){
	ll sum = 0;
	for (int i = 1; i <= N; i++){
		sum += min((ll)N, candidate / i);
	}
	return sum >= K;
}

void pro() {
	ll L = 1, R = (ll)N * N, ans = 0;
	while (L <= R){
		ll mid = (L + R) / 2;
		if (determination(mid)){
			ans = mid;
			R = mid - 1;
		}else L = mid + 1;
	}
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}

File: 강의 자료/02-알고리즘/05~06-이분 탐색/문제별 코드/1300-K 번째 수/solution.py
import sys

n = int(sys.stdin.readline())
k = int(sys.stdin.readline())

def determination(candidate):
    sum = 0
    for i in range(1, n + 1):
        sum += min(n, candidate // i)
    return sum >= k

l, r, ans = 1, n * n, 0
while l <= r:
    mid = (l + r) // 2
    if determination(mid):
        ans = mid
        r = mid - 1
    else:
        l = mid + 1

print(ans)